test/hotspot/jtreg/runtime/Thread/TestAlwaysPreTouchStacks.java

```diff
@@ -108,9 +126,21 @@ public static void main(String[] args) throws InterruptedException, IOException
                     long reserved = Long.parseLong(m.group(1));
                     long committed = Long.parseLong(m.group(2));
                     System.out.println(">>>>> " + line + ": " + reserved + " - " + committed);
-                    if (committed < (reserved / 2)) {
+                    // This is a bit fuzzy: even with PreTouch we don't commit the full range of what NMT counts
+                    // as thread stack. But without pre-touching, the thread stacks would be committed to about 1/5th
+                    // of their reserved size. Requiring them to be committed for over 3/4th shows that pretouch is
+                    // really working.
+                    if ((double)committed < ((double)reserved * 0.75)) {
                         throw new RuntimeException("Expected a higher ratio between stack committed and reserved.");
                     }
+                    // Added sanity tests: we expect our test threads to be still alive when NMT prints its final
+                    // report, so their stacks should dominate the NMT-reported total stack size.
+                    long max_reserved = memoryCeilingMB * 3 * MB;
+                    long min_reserved = memoryCeilingMB * MB;
```

Comment:
How were these boundaries determined? What variance do you see in the reported value when the test runs?

Reply:
>How were these boundaries determined? 

There is a vagueness about reserved total stack size of course - it contains other thread's stack sizes too, and any overhead that the VM may add atop of user-specified thread stack sizes, mainly guard pages.

Min is the logical lower end and the actual number should be greater.
Max is chosen very generously to be much larger. Maybe too large, but I did not want to play whack-the-mole for false positives on platforms I have not tested manually (e.g. larger paged ones), while still excluding obviously insane cases.

On the platforms I tested on variance was not large at all; my 128MB reserved stack size gives me:

linux x86: 133MB
linux aarch64 (4k pages): 154MB
linux x64: 144MB

(The difference between aarch64 and x64 is interesting but I did not dive deeper into that; I assume number of hotspot threads may differ for threads started based on number of cores, but my x64 machine is way beefier than my aarch64-box so the ratio should be reversed)..